File: src/storage/day_store.py
"""
Day-level context (location / weather) stored as JSON per calendar date.

Merge priority: phone > desktop > manual; same rank compares context_updated_at.
"""
from __future__ import annotations

import json
import logging
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Optional

from utils.paths import day_json_relpath

logger = logging.getLogger("diary.storage.day")

CONTEXT_RANK = {"phone": 3, "desktop": 2, "manual": 1, "": 0}


@dataclass
class DayContext:
    date: str = ""
    location: str = ""
    weather: str = ""
    temp_c: Optional[float] = None
    context_source: str = ""
    context_updated_at: str = ""
    updated_at: str = ""


def can_overwrite_context(existing_source: str, new_source: str) -> bool:
    return CONTEXT_RANK.get(new_source, 0) >= CONTEXT_RANK.get(existing_source or "", 0)


def merge_day_context(server: Optional[DayContext], incoming: DayContext) -> DayContext:
    if server is None:
        out = incoming
        if not out.updated_at:
            out.updated_at = out.context_updated_at
        return out
    in_rank = CONTEXT_RANK.get(incoming.context_source or "", 0)
    srv_rank = CONTEXT_RANK.get(server.context_source or "", 0)
    if in_rank > srv_rank or (
        in_rank == srv_rank
        and (incoming.context_updated_at or "") > (server.context_updated_at or "")
    ):
        winner = incoming
    else:
        winner = server
    updated = winner.updated_at or winner.context_updated_at
    return DayContext(
        date=winner.date or incoming.date or server.date,
        location=winner.location,
        weather=winner.weather,
        temp_c=winner.temp_c,
        context_source=winner.context_source,
        context_updated_at=winner.context_updated_at,
        updated_at=updated or incoming.updated_at or server.updated_at,
    )


class DayStore:
    def __init__(self, diary_root: Path) -> None:
        self.diary_root = diary_root
        self.diary_root.mkdir(parents=True, exist_ok=True)

    def path_for(self, entry_date: str) -> Path:
        return self.diary_root / day_json_relpath(entry_date)

    def read(self, entry_date: str) -> Optional[DayContext]:
        path = self.path_for(entry_date)
        if not path.exists():
            return None
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError) as exc:
            logger.warning("Failed to read day context %s: %s", path, exc)
            return None
        temp = data.get("temp_c")
        return DayContext(
            date=data.get("date", entry_date),
            location=data.get("location", ""),
            weather=data.get("weather", ""),
            temp_c=float(temp) if temp is not None else None,
            context_source=data.get("context_source", ""),
            context_updated_at=data.get("context_updated_at", ""),
            updated_at=data.get("updated_at", ""),
        )

    def write(self, ctx: DayContext) -> str:
        """Write day context JSON. Returns relative path from diary root."""
        rel = day_json_relpath(ctx.date)
        path = self.diary_root / rel
        path.parent.mkdir(parents=True, exist_ok=True)
        payload = {
            "date": ctx.date,
            "location": ctx.location,
            "weather": ctx.weather,
            "temp_c": ctx.temp_c,
            "context_source": ctx.context_source,
            "context_updated_at": ctx.context_updated_at,
            "updated_at": ctx.updated_at or ctx.context_updated_at,
        }
        path.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")
        logger.debug("Wrote day context %s", path)
        return rel

    def merge_write(self, incoming: DayContext) -> DayContext:
        existing = self.read(incoming.date)
        merged = merge_day_context(existing, incoming)
        self.write(merged)
        return merged

    def exists(self, entry_date: str) -> bool:
        return self.path_for(entry_date).exists()

File: android/app/src/main/java/com/sparkbox/android/data/DayStore.kt
package com.sparkbox.android.data

import org.json.JSONObject
import java.io.File

private val CONTEXT_RANK = mapOf("phone" to 3, "desktop" to 2, "manual" to 1, "" to 0)

fun mergeDayContext(server: DayContext?, incoming: DayContext): DayContext {
    if (server == null) {
        val updated = incoming.updatedAt.ifBlank { incoming.contextUpdatedAt }
        return incoming.copy(updatedAt = updated)
    }
    val inRank = CONTEXT_RANK[incoming.contextSource] ?: 0
    val srvRank = CONTEXT_RANK[server.contextSource] ?: 0
    val winner = when {
        inRank > srvRank -> incoming
        inRank < srvRank -> server
        (incoming.contextUpdatedAt) > (server.contextUpdatedAt) -> incoming
        else -> server
    }
    val updated = winner.updatedAt.ifBlank { winner.contextUpdatedAt }
    return winner.copy(
        date = winner.date.ifBlank { incoming.date.ifBlank { server.date } },
        updatedAt = updated.ifBlank { incoming.updatedAt.ifBlank { server.updatedAt } },
    )
}

class DayStore(private val diaryRoot: File) {
    init {
        diaryRoot.mkdirs()
    }

    fun pathFor(entryDate: String): File {
        val parts = entryDate.split("-")
        require(parts.size == 3) { "bad date $entryDate" }
        val (y, m, _) = parts
        return File(diaryRoot, "$y/$m/$entryDate.day.json")
    }

    fun read(entryDate: String): DayContext? {
        val f = pathFor(entryDate)
        if (!f.isFile) return null
        return try {
            val json = JSONObject(f.readText(Charsets.UTF_8))
            val temp = if (json.isNull("temp_c")) null else json.optDouble("temp_c")
            DayContext(
                date = json.optString("date", entryDate),
                location = json.optString("location", ""),
                weather = json.optString("weather", ""),
                tempC = temp,
                device = json.optString("device", ""),
                contextSource = json.optString("context_source", ""),
                contextUpdatedAt = json.optString("context_updated_at", ""),
                updatedAt = json.optString("updated_at", ""),
            )
        } catch (_: Exception) {
            null
        }
    }

    fun getOrEmpty(entryDate: String): DayContext =
        read(entryDate) ?: DayContext(date = entryDate)

    fun write(ctx: DayContext) {
        val file = pathFor(ctx.date)
        file.parentFile?.mkdirs()
        val payload = JSONObject()
            .put("date", ctx.date)
            .put("location", ctx.location)
            .put("weather", ctx.weather)
            .put("device", ctx.device)
            .put("context_source", ctx.contextSource)
            .put("context_updated_at", ctx.contextUpdatedAt)
            .put("updated_at", ctx.updatedAt.ifBlank { ctx.contextUpdatedAt })
        if (ctx.tempC != null) payload.put("temp_c", ctx.tempC)
        file.writeText(payload.toString(2), Charsets.UTF_8)
    }

    fun mergeWrite(incoming: DayContext): DayContext {
        val merged = mergeDayContext(read(incoming.date), incoming)
        write(merged)
        return merged
    }

    fun exists(entryDate: String): Boolean = pathFor(entryDate).isFile

    fun listDates(): List<String> {
        if (!diaryRoot.exists()) return emptyList()
        return diaryRoot.walkTopDown()
            .filter { it.isFile && it.name.endsWith(".day.json") }
            .map { it.name.removeSuffix(".day.json") }
            .filter { it.matches(Regex("""\d{4}-\d{2}-\d{2}""")) }
            .toList()
    }
}
